tests/e2e/Services/Projects/ProjectsDevKeys.php

```diff
@@ -0,0 +1,265 @@
+        $this->assertEmpty($response['body']['accessedAt']);
+
+        /** TEST expiry date is required */
+        $res = $this->client->call(Client::METHOD_POST, '/projects/' . $id . '/development-keys', array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+        ], $this->getHeaders()), [
+            'name' => 'Key Test'
+        ]);
+
+        $this->assertEquals(400, $res['headers']['status-code']);
+
+        $data = array_merge($data, [
+            'keyId' => $response['body']['$id'],
+            'secret' => $response['body']['secret']
+        ]);
+
+        return $data;
+    }
+
+
+    /**
+     * @depends testCreateProjectDevKey
+     * @group devKeys
+     */
+    public function testListProjectDevKey($data): array
+    {
+        $id = $data['projectId'] ?? '';
+
+        $response = $this->client->call(Client::METHOD_GET, '/projects/' . $id . '/development-keys', array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+        ], $this->getHeaders()), []);
+
+
+        $this->assertEquals(200, $response['headers']['status-code']);
+        $this->assertEquals(1, $response['body']['total']);
+
+        return $data;
+    }
+
+
+    /**
+     * @depends testCreateProjectDevKey
+     * @group devKeys
+     */
+    public function testGetProjectDevKey($data): array
+    {
+        $id = $data['projectId'] ?? '';
+        $keyId = $data['keyId'] ?? '';
+
+        $response = $this->client->call(Client::METHOD_GET, '/projects/' . $id . '/development-keys/' . $keyId, array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+        ], $this->getHeaders()), []);
+
+        $this->assertEquals(200, $response['headers']['status-code']);
+        $this->assertNotEmpty($response['body']['$id']);
+        $this->assertEquals($keyId, $response['body']['$id']);
+        $this->assertEquals('Key Test', $response['body']['name']);
+        $this->assertNotEmpty($response['body']['secret']);
+        $this->assertArrayHasKey('sdks', $response['body']);
+        $this->assertEmpty($response['body']['sdks']);
+        $this->assertArrayHasKey('accessedAt', $response['body']);
+        $this->assertEmpty($response['body']['accessedAt']);
+
+        /**
+         * Test for FAILURE
+         */
+        $response = $this->client->call(Client::METHOD_GET, '/projects/' . $id . '/development-keys/error', array_merge([
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+        ], $this->getHeaders()), []);
+
+        $this->assertEquals(404, $response['headers']['status-code']);
+
+        return $data;
+    }
+
+    /**
+     * @depends testCreateProject
+     * @group devKeys
+     */
+    public function testNoRateLimitWithDevKey($data): void
```

Comment:
I dont think we have test that rateLimit works, because otherwise tests would start failing after your switch to disabled abuse.


For that reason, I would say to add such a test now that we have noRateLimit test

Reply:
The test for failure below passes a expired dev key for that reason and ge hit rate limit. so I would say the rate limit test works as intended?